floss/language/rust/decode_utf8.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+import sys
+import logging
+import pathlib
+import argparse
+from typing import Any, List, Tuple, Iterable, Optional
+
+import pefile
+
+MIN_STR_LEN = 4
+
+logger = logging.getLogger(__name__)
+
+
+def get_rdata_section(pe: pefile.PE) -> pefile.SectionStructure:
+    for section in pe.sections:
+        if section.Name.startswith(b".rdata\x00"):
+            return section
+
+    raise ValueError("no .rdata section found")
+
+
+def extract_utf8_strings_from_buffer(buf, min_length=MIN_STR_LEN) -> List[List[Any]]:
+    """
+    Extracts UTF-8 strings from a buffer.
+    """
+
+    # Reference: https://en.wikipedia.org/wiki/UTF-8
+
+    character_and_index = []
+    strings = []
+
+    for i in range(0, len(buf)):
+        # for 1 byte
+        if buf[i] & 0x80 == 0x00:
+            character = buf[i].to_bytes(1, "big").decode("utf-8", "ignore")
+            character_and_index.append([character, i, 1])
+
+        # for 2 bytes
+        elif buf[i] & 0xE0 == 0xC0:
+            temp = buf[i] << 8 | buf[i + 1]
+            character = temp.to_bytes(2, "big").decode("utf-8", "ignore")
```

Comment:
why do you use `ignore` here? wouldn't we want to handle the case that invalid UTF-8 data is encountered (and not extract a string there)?

I assume that your algorithm works pretty well, since you've opened the PR, but I can't quite follow how it works. Would you please add some comments explaining the design, and definitely a few test cases that exercise each of the branch arms?

Reply:
Hi, the tests for each branch are in `tests/test_utf8_decoder.py`. Let me know if anything else is required.